tsl/src/fdw/deparse.c

```diff
@@ -834,6 +834,128 @@ deparseSelectStmtForRel(StringInfo buf, PlannerInfo *root, RelOptInfo *rel, List
 	deparseLockingClause(&context);
 }
 
+/*
+ * Construct "SELECT DISTINCT target_list" or "SELECT DISTINCT ON (col1, col..)
+ * target_list" statement to push down the DISTINCT clause to the remote side.
+ *
+ * We only allow references to basic "Vars" or constants in the DISTINCT exprs
+ *
+ * So, "SELECT DISTINCT col1" is fine but "SELECT DISTINCT 2*col1" is not.
+ *
+ * "SELECT DISTINCT col1, 'const1', NULL, col2" which is a mix of column
+ * references and constants is also supported. Everything else is not supported.
+ *
+ * It should be noted that "SELECT DISTINCT col1, col2" will return the same
+ * set of values as "SELECT DISTINCT col2, col1". So nothing additional needs
+ * to be done here as the upper projection will take care of any ordering
+ * between the attributes.
+ *
+ * We also explicitly deparse the distinctClause entries only for the
+ * "DISTINCT ON (col..)" case. For regular DISTINCT the targetlist
+ * deparsing which happens later is good enough
+ */
+static void
+deparseDistinctClause(StringInfo buf, deparse_expr_cxt *context)
+{
+	PlannerInfo *root = context->root;
+	Query *query = root->parse;
+	ListCell *l, *dc_l;
+	bool first = true;
+	bool dc_pushdown_safe = true;
```

Comment:
due to other refactoring you can get rid of dc_pushdown_safe now too and immediately return whenever we detect we cannot pushdown, this removes some nesting in the bottom of the function

Reply:
Ok, done